lib/improver/blending/spatial_weights.py

```diff
@@ -0,0 +1,275 @@
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module to adjust weights spatially based on missing data in input cubes."""
+
+import iris
+import numpy as np
+
+from scipy.ndimage.morphology import distance_transform_edt
+
+from improver.utilities.rescale import rescale
+from improver.utilities.cube_checker import check_cube_coordinates
+
+
+class SpatialWeightsForMissingData(object):
+    """
+    Plugin for adjusting weights spatially based on missing data in the input
+    cube that will be collapsed using these weights.
+
+    The plugin does the following steps:
+
+        1. Creates an initial spatial weights basked on the mask in the
+           input cube giving zero weight to where there is masked data and
+           a weight of 1 where there is valid data.
+        2. Make these weights fuzzy by smoothing the boundary between where
+           there is valid data and no valid data. This keeps areas of zero
+           weight, but reduces nearby grid points with a weight of 1 depending
+           on how far they are from a grid point with a weight of zero. The
+           range of the effect is controlled by the supplied fuzzy_length
+        3. Multiplies the fuzzy spatial weights by the one_dimensional weights
+           supplied to the plugin.
+        4. Normalises the weights along the coordinatethat will be collapsed
+           when blending is carried out using these weights.
+    """
+
+    def __init__(self, fuzzy_length=10):
+        """
+        Initialise class.
+        Keyword Args:
+            fuzzy_length (int or float):
+                The length in terms of number of grid squares, over which the
+                weights from the input data mask are smoothed.
+        """
+        self.fuzzy_length = fuzzy_length
+
+    def __repr__(self):
+        """Represent the configured plugin instance as a string."""
+        result = ('<SpatialWeightsForMissingData: fuzzy_length: {}>'.format(
+                      self.fuzzy_length))
+        return result
+
+    @staticmethod
+    def create_initial_weights_from_mask(cube):
+        """
+        Generate a cube with weights generated from the mask of the input cube.
+        Where the data is masked we set the weight to zero, otherwise the
+        weight is one.
+
+        Args:
+            cube (iris.cube.Cube):
+                A cube containing the data we want to collapse by doing a
+                weighted blend along a given cordinate.
+        Return:
+            masked_weights_cube (iris.cube.Cube):
```

Comment:
The first time I read this I assumed that the result was a masked cube. Would it be better to call this weights_from_mask